### PR TITLE
Adds streaming option to generate

### LINKED ISSUE #1420: Stream option
Add a `stream=True` option to `generate`

### PR DESCRIPTION
**Pull request by rasbt**
Fixes #1420

### CODE REVIEW

**Comment by rasbt** [comment]
Is this what you were looking for @williamFalcon ?

**Comment by williamFalcon** [comment]
nice! 

can you test it with this Studio
https://lightning.ai/lightning-ai/studios/deploy-a-private-llama-3-8b-api

**Comment by rasbt** [comment]
It should work now @williamFalcon . 

I modeled the adjusted `litgpt serve`, which optionally uses the `stream=True` option in `generate`  after [this example](https://github.com/Lightning-AI/LitServe/blob/3eb31aac3af89ba37cfa9698df9097332677416e/tests/e2e/default_batched_streaming.py#L20) in `LitServe`. Since I am not super familiar with LitServe, could you double-check that this is indeed the correct use @aniketmaurya and @lantiga ?

File: litgpt/deploy/serve.py [hunk 1]
```diff
@@ -106,6 +123,42 @@ def encode_response(self, output: torch.Tensor) -> Dict[str, Any]:
...
+
+    def predict(self, inputs: torch.Tensor) -> Any:
+        # Run the model on the input and return the output.
+        prompt_length = inputs.size(0)
+        max_returned_tokens = prompt_length + self.max_new_tokens
+
+        for block in self.model.transformer.h:
+            block.attn.kv_cache.reset_parameters()
+
+        yield generate(
```
Comment (aniketmaurya) [inline litgpt/deploy/serve.py:147]: should be `yield from generate`

Comment (aniketmaurya) [reply litgpt/deploy/serve.py:147]: right now, a generator is yielded and encode_response doesn't iterate through the response of `generate`.

```py
def encode_response(self, output_stream):
        ### Output stream here will basically behave like [generator(...)]
        for outputs in output_stream:

```

File: litgpt/deploy/serve.py [hunk 2]
```diff
@@ -106,6 +123,42 @@ def encode_response(self, output: torch.Tensor) -> Dict[str, Any]:
...
+                max_returned_tokens,
+                temperature=self.temperature,
+                top_k=self.top_k,
+                top_p=self.top_p,
+                eos_id=self.tokenizer.eos_id
+                )
+
+    def encode_response(self, output_stream):
+        for outputs in output_stream:
+            yield [json.dumps({"output": self.tokenizer.decode(output)}) for output in outputs]
```
Comment (aniketmaurya) [inline litgpt/deploy/serve.py:159]: LitServe can now do the https://github.com/Lightning-AI/LitServe/issues/84 so we can remove it from here.

File: litgpt/deploy/serve.py [hunk 3]
```diff
@@ -106,6 +123,42 @@ def encode_response(self, output: torch.Tensor) -> Dict[str, Any]:
...
+                self.model,
+                inputs,
+                max_returned_tokens,
+                temperature=self.temperature,
+                top_k=self.top_k,
+                top_p=self.top_p,
+                eos_id=self.tokenizer.eos_id
+                )
+
+    def encode_response(self, output_stream):
```
Comment (aniketmaurya) [inline litgpt/deploy/serve.py:157]: looping over outputs [example here](https://github.com/Lightning-AI/LitServe/blob/3eb31aac3af89ba37cfa9698df9097332677416e/tests/e2e/default_batched_streaming.py#L20) is for streaming with batching but this implementation doesn't contain batching so we need to return `yield {"output": self.tokenizer.decode(output)}`

File: litgpt/deploy/serve.py (hunk 2 above)
Comment (rasbt) [reply litgpt/deploy/serve.py:159]: Oh good that you bring this up ... I remember now that I pinned the LitServe version, which maybe explains the need for this 😅
https://github.com/Lightning-AI/litgpt/blob/5dabf5f6be507e4da4d7be69813e6f31e799aefd/pyproject.toml#L41

Comment (rasbt) [reply litgpt/deploy/serve.py:159]: I think this has not been in the LitServe release yet right? So we would need to revisit this later in LitGPT.

**Comment by rasbt** [comment]
Many thanks for the feedback @aniketmaurya ! Does the current code look better?

**Review by carmocca — COMMENTED** [review]
Why not import and use the chat.py generation implementation which already implements streaming? Instead of modifying the generate.py file

File: litgpt/generate/base.py [hunk 4]
```diff
@@ -78,7 +78,8 @@ def generate(
     top_k: Optional[int] = None,
     top_p: float = 1.0,
     eos_id: Optional[int] = None,
-) -> torch.Tensor:
+    stream: bool = False,
```
Comment (aniketmaurya) [inline litgpt/generate/base.py:81]: the above StreamLitAPI class seem to be missing `generate(..., stream=True)` argument.

File: litgpt/deploy/serve.py [hunk 5]
```diff
@@ -106,6 +123,41 @@ def encode_response(self, output: torch.Tensor) -> Dict[str, Any]:
...
+                inputs,
+                max_returned_tokens,
+                temperature=self.temperature,
+                top_k=self.top_k,
+                top_p=self.top_p,
+                eos_id=self.tokenizer.eos_id
+            )
+
+    def encode_response(self, output):
+        yield {"output": self.tokenizer.decode(next(output))}
```
Comment (aniketmaurya) [inline litgpt/deploy/serve.py:158]: the `output` argument here would be a generator and is of arbitrary size. we need to iterate over output and yield each out the values.

```py
def encode_response(self, output):
    for out in output:
        yield {"output": self.tokenizer.decode(out)}
```

**Review by aniketmaurya — CHANGES_REQUESTED** [review]
LGTM! added few comments. 

Also please take a latest pull of LitServe, last week we merged some PR which affects streaming endpoint.

**Comment by rasbt** [comment]
@carmocca That's a good question, I am not sure why I was so fixated on `generate`. Let me see (in a separate PR) if I can make this serving example work with `chat`, then we don't need to modify `generate`.


Thanks for the feedback, @aniketmaurya !

> Also please take a latest pull of LitServe, last week we merged some PR which affects streaming.

I could try that internally for testing but we'd probably have to wait until the next release. But could you ping me when the latest LitServe comes out so that I can bump the version in LitGPT?

**Comment by rasbt** [comment]
Thanks for all your help with LitServe so far @aniketmaurya . It's probably something stupid on my part, but with the recent LitServe version from GitHub main somehow it's still not working right 

I.e. with

```
litgpt  serve --checkpoint_dir checkpoints/EleutherAI/pythia-1b --stream false

response = requests.post(
        "http://127.0.0.1:8000/predict",
         json={"prompt": "Fix typos in the following sentence: Exampel
         input"}
 )
```

The `response.content` comes out as 

```
--stream true

b'{"output": " file"}{"output": ":"}{"output": " /"}{"output": "usr"}{"output": "/"}{"output": "share"}{"output": "/"}{"output": "X"}{"output": "11"}{"output": "/"}{"output": "x"}{"output": "org"}{"output": "."}{"output": "conf"}{"output": "."}{"output": "d"}{"output": "/"}{"output": "x"}{"output": "org"}{"output": "."}{"output": "conf"}{"output": "."}{"output": "7"}{"output": "."}{"output": "idx"}{"output": "."}{"output": "gz"}{"output": ":"}{"output": " error"}{"output": " opening"}{"output": " /"}{"output": "usr"}{"output": "/"}{"output": "share"}{"output": "/"}{"output": "X"}{"output": "11"}{"output": "/"}{"output": "x"}{"output": "org"}{"output": "."}{"output": "conf"}{"output": "."}{"output": "d"}{"output": "/"}{"output": "x"}{"output": "org"}{"output": "."}{"output": "conf"}{"output": "."}'
```

whereas `--stream false` works as expected via LitServe (pls ignore the bad Pythia response, I just used a small model here for demoing:

```
b'{"output":"Fix typos in the following sentence: Exampel input.\\n\\nThis book is now available as a Kindle app.\\n\\nPlease go to the print page to view the rest of this book.\\n\\n## **Contents**\\n\\n**1** The Problem\\n\\n**2** How We Lear"}'
```

I think it's something in my LitServe implementation because the `generate` with `stream=True` works as intended (e.g., via `litgpt generate base --checkpoint_dir checkpoints/EleutherAI/pythia-1b --stream true`)

If you have some time to take another look, that'd be really appreciated.

**Comment by aniketmaurya** [comment]
Hi @rasbt , we will have to iterate through the response in streaming mode like:

```py

import requests

url = "http://127.0.0.1:8000/predict"

resp = requests.post(url, json={"prompt": "Fix typos in the following sentence: Exampel input"}, headers=None,
                     stream=True)
for line in resp.iter_lines():
    if line:
        print(line.decode("utf-8"))
```


when we don't iterate and use response.content, requests will concat everything in a bytestring.

**Comment by rasbt** [comment]
That makes sense! Thanks for your patience here @aniketmaurya , I know almost nothing about web stuff and serving. The code you shared does make sense, but yeah, it's still only generating a single token. Something must still be incorrect somewhere. 

However, I also wonder if there's maybe a LitServe issue. I tried a simpler example:

```python
import litserve as ls
import json


class SimpleStreamAPI(ls.LitAPI):
    def setup(self, device) -> None:
        self.model = lambda x: str(x)

    def decode_request(self, request):
        return request["input"]

    def predict(self, x):
        for i in range(10):
            yield self.model(i)

    def encode_response(self, output_stream):
        for outputs in output_stream:
            yield [json.dumps({"output": output}) for output in outputs]


if __name__ == "__main__":
    server = ls.LitServer(SimpleStreamAPI(), stream=True)
    server.run(port=8000)
```

and then via

```python
import requests

url = "http://127.0.0.1:8000/predict"

resp = requests.post(url, json={"input": "1, 2, 3"}, headers=None, stream=True)
for line in resp.iter_lines():
    if line:
        print(line.decode("utf-8"))
```

it just returns `["{\"output\": \"0\"}"]`, which doesn't seem right. I would expect it to return 10 things due to the for-loop in predict, or am misunderstanding this?

**Comment by rasbt** [comment]
Ok great, this works flawlessly with the new LitServe version now:

```python
url = "http://127.0.0.1:8000/predict"

resp = requests.post(url, json={"prompt": "Hello world"}, headers=None, stream=True)
for line in resp.iter_lines():
    if line:
        print(line.decode("utf-8"))
```

```
{"output": "\n"}{"output": "\n"}{"output": "It"}{"output": " seems"}{"output": " that"}{"output": " I"}{"output": " can"}{"output": "'t"}{"output": " load"}{"output": " my"}{"output": " php"}{"output": " page"}{"output": " from"}{"output": " my"}{"output": " html"}{"output": " file"}{"output": "."}{"output": "\n"}{"output": "\n"}{"output": "A"}{"output": ":"}{"output": "\n"}{"output": "\n"}{"output": "1"}{"output": ")"}{"output": " Check"}{"output": " this"}{"output": " link"}{"output": ","}{"output": " it"}{"output": " might"}{"output": " solve"}{"output": " your"}{"output": " problem"}{"output": "."}{"output": "\n"}{"output": "2"}{"output": ")"}{"output": "Check"}{"output": " this"}{"output": " article"}{"output": "."}{"output": "\n"}{"output": "3"}{"output": ")"}{"output": " check"}{"output": " your"}{"output": " php"}{"output": " file"}{"output": " path"}
```

**Comment by rasbt** [comment]
I now also got it to work with the chat version via #1426. I guess we can close this one then and leave the base generate function untouched.

Or alternatively, the question is if we should replace the `generate` function `chat/base.py` with this one here that can handle both streaming and non-streaming @carmocca ? In the end, it would be less code to maintain (only one instead of two generate functions).

**Comment by carmocca** [comment]
I didn't merge them before because the logic to check for the stop tokens is not as performant as simply generating a fixed number of tokens. They could probably still be reconcilliated with some if-elses but code simplicity over deduplication was desirable before

**Comment by rasbt** [comment]
Makes sense and no worries. I'd say let's keep it as is. I got the LitServe streaming example to work after all, so mission accomplished 😊

**Pull request closed without merging**